### ISSUE TITLE
DevWorkspace Che routing should propagate newer hosts rules

### ISSUE THREAD

**Issue by sleshchenko** [eclipse-che/che#19858]
### Describe the bug
A PR that fixes Che DWR since wildcard certificate with one * does not work with sub-sub domains. And it works fine for newly created devworkspaces, but existing ones stick to their old invalid hosts.
DWCO should always make sure that DevWorkspace uses the latest hosts rules.

### Che version

 - [ ] latest
 - [ ] nightly
 - [ ] other: please specify

### Steps to reproduce
1. Deploy Che + devWorkspace where DWCO is built from changes before PR
2. Create DevWorkspace.
3. Update DWCO to latest :ci
4. It's not possible to open DevWorkspace due invalid TLS certificate.

### Expected behavior
Newer DWCO updated devworkspace hosts and TLS certificate issue is fixed.

P.S. That's OK if DWCO require DevWorkspace to restart, but the current approach means - such devworkspaces must be recreated with possible information losing.

### Runtime
   - [ ] kubernetes (include output of `kubectl version`)
   - [ ] Openshift (include output of `oc version`)
   - [ ] minikube (include output of `minikube version` and `kubectl version`)
   - [ ] minishift (include output of `minishift version` and `oc version`)
   - [ ] docker-desktop + K8S (include output of `docker version` and `kubectl version`)
   - [ ] other: (please specify)

### Installation method
   - [ ] chectl
     * provide a full command that was used to deploy Eclipse Che (including the output)
     * provide an output of `chectl version` command
   - [ ] OperatorHub
   - [ ] I don't know

### Environment
   - [ ] my computer
       - [ ] Windows
       - [ ] Linux
       - [ ] macOS
   - [ ] Cloud
       - [ ] Amazon
       - [ ] Azure
       - [ ] GCE
       - [ ] other (please specify)
   - [ ] Dev Sandbox (workspaces.openshift.com)
   - [ ] other: please specify

**Comment by l0rd** [comment]
@sleshchenko with DW I would like to start deprecating sub-domain routing (and only support path routing). Is this issue related to DWR supporting sub-domains?

**Comment by sleshchenko** [comment]
I believe it's also actual for path routing, but I let experts answer for sure @metlos @sparkoo

**Comment by metlos** [comment]
I think we actually should not change this behavior.

In my humble understanding, this feature request asks for changing the exposed URLs of a running workspace. IMHO, that is not a good idea. As pointed out in the description, the workspaces start up fine when restarted, which IMHO is how it should behave, because IDEs and any other tooling can count on the URLs being stable during the lifetime of the workspace.

**Comment by sleshchenko** [comment]
> In my humble understanding, this feature request asks for changing the exposed URLs of a running workspace

@metlos from the PR description

> P.S. That's OK if DWCO requires DevWorkspace to restart, but the current approach means - such devworkspaces must be recreated with possible information losing.

maybe it's outdated behavior.
Or do you expect DWO to recreate DWR on each workspace start?

**Comment by metlos** [comment]
> Or do you expect DWO to recreate DWR on each workspace start?

Ok, I see your point - DWRs are not deleted when a workspace is stopped and therefore all the ingresses and services remain in place even if the devworkspace deployment is scaled down.

Frankly, I don't see how we could be 100% bullet-proof here. We've encountered the problem because DWCO was buggy and created invalid host/tls pairs.

I also consider the current behavior of keeping the DWR and associated resources even if the workspace is stopped a good design.

So maybe this boils down to remembering to put a bullet point in the release notes of the (DW)CO when such changes happen - this would be fixed by deleting the DWRs of the workspaces, which would force their re-creation and re-creation of the ingresses/routes/services without touching the workspace deployment or devworkspace CR itself and maybe reloading the workspace in the browser.

Looking at the code again, I actually don't know the reason why Host in RouteSpec is ignored when reconciling. This is coming from DWO (https://github.com/devfile/devworkspace-operator/blob/main/controllers/controller/devworkspacerouting/sync_routes.go#L33). Maybe this is to tolerate the situation where the hostname is determined by OpenShift itself? Not sure...

On the other hand DWCO has its own problem with this, because its hostname creation algorithm is not stable because it depends on a map iteration order (by its own recognition in the comments :) ). We should fix that, but I've created a separate issue for that - https://github.com/eclipse/che/issues/20176
